go/interpreter/lfvm/instructions.go

```diff
@@ -1119,44 +1127,64 @@ func opCall(c *context) {
 		cost += tosca.Gas(params.CallStipend)
 	}
 
-	// If we are in static mode, recursive calls are to be treated like
-	// static calls. This is a consequence of the unification of the
-	// interpreter interfaces of EVMC and Geth.
-	// This problem was encountered in block 58413779, transaction 7.
-	kind := tosca.Call
-	if c.params.Static {
-		kind = tosca.StaticCall
-
-		// In a static call, no value must be transferred.
+	if kind == tosca.Call || kind == tosca.CallCode {
+		// Check that the caller has enough balance to transfer the requested value.
 		if !value.IsZero() {
-			c.SignalError(errWriteProtection)
-			return
+			balance := c.context.GetBalance(c.params.Recipient)
+			balanceU256 := new(uint256.Int).SetBytes32(balance[:])
+			if balanceU256.Lt(value) {
+				c.stack.pushEmpty().Clear()
+				c.return_data = nil
+				c.gas += cost // the gas send to the nested contract is returned
+				return
+			}
 		}
 	}
-
-	// Get the arguments from the memory.
+	// Get arguments from the memory.
 	args := c.memory.GetSlice(inOffset.Uint64(), inSize.Uint64())
 
-	// Check that the caller has enough balance to transfer the requested value.
-	if !value.IsZero() {
-		balance := c.context.GetBalance(c.params.Recipient)
-		balanceU256 := new(uint256.Int).SetBytes32(balance[:])
-		if balanceU256.Lt(value) {
-			c.stack.pushEmpty().Clear()
-			c.return_data = nil
-			c.gas += cost // the gas send to the nested contract is returned
-			return
+	callParams := tosca.CallParameters{}
+	subKind := kind
+	switch kind {
```

Comment:
nit:
- 'c.params.Static' check can be done in the opCall function, no need for subkind in this function.
- Input and gas are common to all cases, do once. 
- Value can be common, and set to c.params.Value for DelegateCall

This reduces decision-making in the switch, and it is easier to read, sender, recipient and code address.